src/ring_theory/graded_algebra/basic.lean

```diff
@@ -0,0 +1,57 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser, Kevin Buzzard, Jujian Zhang
+-/
+import algebra.direct_sum.algebra
+import algebra.direct_sum.internal
+
+/-! # Typeclass for graded ring
+For definition of an `R`-algebra `A` being graded by `𝒜 : ι → submodule R A`, see doc string of
+`graded_algebra`.
+
+- `graded_algebra.decompose : A → ⨁ i, 𝒜 i` and `graded_algebra.recompose : ⨁ i, 𝒜 i → A` are
+the algebra isomorphism between `A` and `⨁ i, 𝒜 i` if `A` is graded by `𝒜`.
+-/
+
+open_locale direct_sum big_operators
+
+section graded_ring
+variables (R A : Type*) [comm_semiring R] [ring A] [algebra R A]
+  {ι : Type*} (𝒜 : ι → submodule R A)
+  [decidable_eq ι] [add_comm_monoid ι]
```

Comment:
I'd lean towards putting the types first:
```suggestion
variables {ι R A : Type*}
variables [decidable_eq ι] [add_comm_monoid ι] [comm_semiring R] [ring A] [algebra R A]
variables (𝒜 : ι → submodule R A)
```
Note also that you can make `R` and `A` explicit, as both are implied by `𝒜`